### ISSUE TITLE
Stencils for operators

### ISSUE THREAD

**Issue by thomasmelvin** [stfc/PSyclone#1103]
LFRic [#2446](https://code.metoffice.gov.uk/trac/lfric/ticket/2446) introduces a new kernel that requires access to operators in a stencil. In order to correctly address these it needs to pass in an array of 2D cell id's for the stencil. This currently isn't supported in lfric or psyclone.

The psykal-lite code (contained in `invoke_helmholtz_operator_kernel_type`) is 

      do cell=1,mesh%get_last_edge_cell()
        !
        ! Populate cell_stencil array used for operators
        ! (this is the id of each cell in the stencil)
        cell_stencil(1) = cell
        do i = 1,4
          cell_stencil(i+1) = mesh%get_cell_next(i, cell)
        end do        

        CALL helmholtz_operator_code(cell_stencil, nlayers, &
                                 ...
                                  )
    end do

**Comment by arporter** [comment]
Thinking about the kernel metadata, will it be sufficient to extend what we currently have for fields and just permit operator arguments to have STENCIL specified? Is the set of stencils (e.g. region, cross2d) the same in both cases?

In terms of implementation, is there any plan/need to implement LFRic infrastructure support for this new stencil or should PSyclone just generate the code illustrated above? If so, we would need rules linking stencil type to the dimensionality/extents of the `cell_stencil` array.

**Comment by thomasmelvin** [comment]
For the metadata I think you are correct we just need to extend our current STENCIL to operator

For the code I think we need to implement new LFRic infrastructure, there are a number of reasons for this:

1.  The example code only creates a depth one cross stencil, greater depths or region stencils are more complicated and so I think this code should not be generated, but could be easily extended from the the existing lfric stencil code
2. My understanding is that having the getters inside the openmp loop is not good for performance and so we would probably want to move this outside of the horizontal loop and get the entire stencil in one go
3.  At the moment operators are computed to the depth 1 halo, but never have the halo updated. If we want a stencil of greater depth, or to  apply to continuous fields then we would need the operators at greater depth into the halo (which isn't currently supported) so we need to decide the best way of doing that. Note this does not apply to this ticket that only loops over owned cells and only requires a depth 1 halo

**Comment by rupertford** [comment]
We currently have check within PSyclone that we conform to your point 3, so, in the short term we'd need to update this check to make it work correctly when there are operator stencils.

**Comment by andrewcoughtrie** [comment]
@thomasmelvin am I right in thinking you need a stencil of cell IDs rather than dof IDs for this? If so I think we should be able to add a function to the LFRic stencils class that returns that in the same way we do for stencils at the moment. The cell IDs are already used in the stencil code in their creation anyway so it should be simple to extend.

**Comment by thomasmelvin** [comment]
@andrewcoughtrie That's correct and was along the lines of what I was thinking of for the infrastructure required

**Comment by andrewcoughtrie** [comment]
I've opened a ticket to implement a stencil of cell IDs in LFRic infrastructure [#2532](https://code.metoffice.gov.uk/trac/lfric/ticket/2532)

**Comment by stevemullerworth** [comment]
I'm currently reviewing the related LFRic [ticket #2446](https://code.metoffice.gov.uk/trac/lfric/ticket/2446)

The code is currently written as follows, and the kernel size is hard-wired in the code.

`CALL helmholtz_operator_code(cell_stencil, nlayers, &`

It seems to me the arguments should include the stencil size. Do people concur?

`CALL helmholtz_operator_code(cell_stencil_size, cell_stencil, nlayers, &`

Note order of arguments matches the use of stencils for individual fields, where size is put first.

**Comment by andrewcoughtrie** [comment]
If the stencil depth is going can be anything other than 1 then yes, to be able to loop over the stencil correctly in the kernel the length of the stencil array would be needed (currently hard coded in the helmholtz operator kernel). How it is used in the kernel suggests it may never be greater than 1 in this case but i'd say it should be a general purpose implementation allowing any depth.

**Comment by thomasmelvin** [comment]
I'd agree, we should pass the stencil size in. Currently the operators wouldn't be computed for a greater than depth 1 halo but that is another problem
